File: problem-solving/_lc-120-medium-triangle/2-bottom-up.py
"""
https://leetcode.com/problems/triangle/

Complexity
----------
Time: O(n*n)
Space: O(n*n)
"""

class Solution:
    def minimumTotal(self, triangle: List[List[int]]) -> int:
        dp = [[0] * len(triangle[r]) for r in range(len(triangle))]
        dp[0][0] = triangle[0][0]
        for r in range(1, len(triangle)):
            for c in range(len(triangle[r])):
                if c == 0:
                    dp[r][c] = triangle[r][c] + dp[r - 1][c]
                elif c == len(triangle[r]) - 1:
                    dp[r][c] = triangle[r][c] + dp[r - 1][c - 1]
                else:
                    dp[r][c] = triangle[r][c] + min(dp[r - 1][c], dp[r - 1][c - 1])
        return min(dp[-1])

File: problem-solving/_lc-123-hard-best-time-to-buy-and-sell-stock-iii/2-bottom-up.go
/*
https://leetcode.com/problems/best-time-to-buy-and-sell-stock-iii/

Complexity
----------
Time: O(n)
Space: O(n)

Reference
---------
Approach-3: https://leetcode.com/problems/best-time-to-buy-and-sell-stock-iii/solutions/796990/c-worst-to-best-solution-explained-for-dummies-like-me/
*/

func maxProfit(prices []int) int {
    buy1, buy2 := math.MaxInt32, math.MaxInt32
    profit1, profit2 := math.MinInt32, math.MinInt32
    for i := 0; i < len(prices); i++ {
        buy1 = min(buy1, prices[i])
        profit1 = max(profit1, prices[i] - buy1)
        buy2 = min(buy2, prices[i] - profit1)
        profit2 = max(profit2, prices[i] - buy2)
    }
    return profit2
}

func min(x, y int) int {
    if x < y {
        return x
    }
    return y
}

func max(x, y int) int {
    if x > y {
        return x
    }
    return y
}
